### ISSUE TITLE
How Should an App Determine a User's Name?

### ISSUE THREAD

**Issue by sandhawke** [solid/solid#74]
How can an app find out the display name of the current user?

How can app find out the name of another user, given their WebID?

PROPOSAL-1: (for illustration purposes, showing what a solution might look like)
* Every WebID document MUST include at least one foaf:name for that WebID.   It doesn't have to be their legal name or anything -- it's just the name they want displayed for them.   
* The profile MUST NOT include more than one foaf:name with the same language tag.  That is, there can be multiple names, but they must all have different language tags.
* Applications SHOULD flag as broken any WebID that doesn't have at least one foaf:name triple

The floor is open for other proposals.....

**Comment by bblfish** [comment]
I think those are reasonable as SHOULDs. 

Given that there is a lot of foaf out there already, we have to work with what is out there too. So the other answer is to ask Designers if there is a good way to deal with multiple names.

Perhaps also someone with a LinkedData crawler can tell us what the situation on the ground is.

**Comment by sandhawke** [comment]
Can you give me a possible reason for omitting a foaf:name, as an example of where a SHOULD would differ from a MUST?

**Comment by bblfish** [comment]
I am not saying one should omit it, I am saying if there are more the UI needs to find a way to show that there are more. There may be good ways to do that.

**Comment by dmitrizagidulin** [comment]
Keeping in mind that the initial proposal was merely for illustration purposes --

I am strongly against requiring `foaf:name` in the WebID profile (heh, especially given your own views on it, in issue #35). 

PROPOSAL-2:

1. A preferred username SHOULD be stored in a user's private Preferences file (determined at signup). Vocab/term tbd (I'll propose some concrete ones when I get a few mins to research). An app MUST look in the preferences first, to see how it should address the user.
2. If the user does not have a preferred name stored, or if the app doesn't have access to the preferences file, the app SHOULD fall back on a hierarchy (tbd) of commonly occurring names:
  - possibly various `foaf:` names and nicknames / usernames
  - fall back on the account name (derived from the WebID, if on a multi-user system like `databox.me`), or on the full WebID itself
3. This cascading logic should be wrapped in a client library (something like Solid.js), so that the app developer can just call `profile.name()`, and not have to re-implement it in every app.

**Comment by sandhawke** [comment]
My goal here is to make the application logic as simple as possible.     I appreciate that it can be encapsulate in libraries, but there are going to be lots of libraries, and some people who don't use libraries.    I feel like a lot of this is unnecessarily complexity.     But maybe it's needed....

ISSUE-74-1: What is the first choice predicate?   foaf:name or something else?

ISSUE-74-2: Do apps have to look anywhere other than the WebID document for other people's names?

ISSUE-74-3: Do apps have to tlook anywhere other than the WebID document for the user's name?

ISSUE-74-4: What if users have multiple names in one language?    Do we allow that?   How do we select among them for display?

ISSUE-74-5: Do we allow for multipart names (first/last, given/family)?  If so, how?

ISSUE-74-6: How do we handle people who want to be "anonymous"?   Do we just display their domain name?    Do we force them to pick some display name?   Do we have something like slashdot's "Anonymous Coward" default name, even though in our case their domain name is available.

ISSUE-74-7: How do we handle impersonation?   If I set my foaf:name to Tim Berners-Lee, will I start to get messages from people trying to write to him?

ISSUE-74-8: Do we have multiple names for different contexts?   Long, short, formal, informal, with and without titles, ...

Procedurally, following issue #72 one can imagine a "User Names" spec, and the above 8 issues are issues on that spec.

**Comment by sandhawke** [comment]
solid:name   ?   :-)

**Comment by dmitrizagidulin** [comment]
Further thoughts on this, after a conversation with Sandro.

We're trying to solve the name problem for *two* different use cases. 

**Use case A:** what should an app call the *current user*?

This is to provide a more friendly user experience, and to help the user tell apart their various logins and accounts. Think of the typical username and user picture/icon on the upper right nav bar, on various social sites.

**Use case B:** what should the app call *other users*/webId profiles that the current user cares about? 

This is more for the various social apps that involve other people, like Contact managers / people pickers, calendaring apps, chat apps, etc. 
When your app is showing you a list of who is invited to your meeting, it might be nice to have shorter names for them than their full WebID urls.

These two use cases have slightly different requirements, and should be implemented slightly differently, I think.

**A - current user** properties:

- can and should be private to the user (and so, live in preferences). a preferred username and icon type of thing, to show in the nav bar
- does not have to be globally unique. If your name is Joe, you can just appear as Joe in your own nav bar, you know who you are 
- can be the account name, if no preferred name is provider. or even the truncated webid (something like 'alice.databox.m...', with the assumption that the UI will hyperlink it to the full webid url

**B - other people** properties:

- has to balance the user's convenience, versus other people's desire to be anonymous 
- should be informed by those people's preferred names in their public profiles. (but not required)
- can always fall back on the acct name or truncated webid url as above
- should be able to be overriden by the current user, for their own apps. That is, even though my friend's preferred username is 'iLuvCars99', I should be able to alias them as 'Josh' to help me remember who they are, when they show up on my meeting's invite list. obviously these aliases should be kept in my private preference files. Current apps like Adium and various Twitter clients etc, let you do this already.

**Comment by mansoure** [comment]
The user has an option in the signup to upload a photo for his WebID account. 
Why not also the user gets an option to give a foaf:name, which might be different from the "user name" of his account? For example, the user name is essam27.databox.me and the user would like to use Essam Mansour as his foaf:name .

**Comment by timbl** [comment]
A)   What you call the currect user depends on the context a lot, very app dependent

- In "logged in as"  status button ?     foaf:nick is good

- In lists  of authorship etc

- "you" is often acceptable in ACLs, as in "you have complete control"   
owner: you

- Never "me"   <--   personal rant 

B) Propose for the name a user is listed with eg as an author or contributor or owner or in ACL:.

1) foaf:name if available.
2) vcard:fn  if available.
3) foaf:nick
4) an attempt to make something from the URI (which is generally very hairy)


NOT solid:name

**Comment by nicola** [comment]
I have once written something of this kind in my app, to avoid the problems ^

```javascript
var name = person.nick || person.nickname || person.givenName || person.name || 'User'
```

**Comment by ghanemabdo** [comment]
I think there is a need for introducing a new term "foaf:displayName" as I feel it's a very common problem we're trying to work it around. the displayName is the public name the user wants to appear with. It's like "Hey friend, your name is ... but how'd you like me to call/see you?".

My proposal is simply:
1) Registering a new WebID, the user SHOULD enter foaf:name or foaf:nickname. foaf:displayName is copied by DEFAULT from foaf:name then foaf:nickname whatever available.
2) The user has the option to change the display name to anything other than the DEFAULT.
3) Having foaf:name and foaf:nickname not entered (which means foaf:displayName is empty):
    - foaf:displayName MUST be retrieved from user's WebID OR
    - foaf:displayName  MUST be set to "Anonymous User", "User", ...etc

My point is that foaf:displayName MUST be set to some value that the user has the option to set him/herself not guessed from other parameters. The existence of this field will simplify the retrieval of user's name (and probably it can be generalized to more fields) instead of having several fallback cases which actually pushes the problem to the application level instead of solving it by design.

**Comment by sandhawke** [comment]
@ghanemabdo I'm not seeing how foafkdisplayName would be different from foaf:name.     If you look at the spec for foaf:name, what would you have be different for displayName?

**Comment by sandhawke** [comment]
@nicola do you want to link to your code that does this?

**Comment by ghanemabdo** [comment]
@sandhawke 
foaf:name => this is my actual name that should appear on my profile when I log in to any app (have the permission to access my personal info) and it's optional. So, from apps perspective, reading data from the logged in user's profile, foaf:name should be read not foaf:displayName.

foaf:displayName => this is my public name that others see me with (does not have permissions to access my personal info) and it's mandatory. Apps reading data from other users' profile docs should seek foaf:displayName as it's guaranteed to exist and has public permission. For example, a chat app populating the list of my friends should seek their foaf:displayName while in the profile bar of the app should display foaf:name retrieved from foaf:name from my profile doc.

My argument is that both points of view are right, user name should be kept private and also there must be a name for the user in the profile doc that everybody can retrieve. The proposal above should be a reasonable solution rather than leaving the decision to the application to decide which way to go with.

**Comment by sandhawke** [comment]
I think I see the difference you're making.   I'd suggest that if we're using foaf:name for one of them (and there are reasons not to use foaf, I realize) then foaf:name is already defined to be the name by which others know you.   So if we're making a second one, it should be the different predicate, like selfName.     I'm not sure I really see it as important though.    When would a user really want it a different private name and public name?      I see different names in different groups as more compelling, and still don't think we need that yet.

**Comment by kidehen** [comment]
@timbl -- 
+1 in regards to it should never be solid:name, under any circumstances. What about leveraging schema.org terms? Net effect:

1) foaf:name if available.
2) vcard:fn if available.
3) [schema:name ](http://schema.org/name) if available 
4) foaf:nick

**Comment by sandhawke** [comment]
I assume the rationale is legacy data?     I don't see why supporting legacy vocabularies is useful here, since other protocols we're requiring are entirely new, so it's unlikely legacy data would be encountered in a useful way.

**Comment by bblfish** [comment]
> I don't see why supporting legacy vocabularies is useful here, since other protocols we're requiring are entirely new, so it's unlikely legacy data would be encountered in a useful way.

HTTP is SoLiD-- , and there is a lot of foaf and vcard out there on SoLiD--. So the protocol is already wildely used, and the data widely distributed.

You can't actually avoid messy data in an open world. We can and should have guidelines about which vocabularies we use but our apps should work with existing data (so that we don't need to start from scratch - foaf is at the center of the linked data cloud http://lod-cloud.net/ ),  and so we need to think about how to use that data and how to display it propertly. So there we can look have competitions to work out who does it best.

We can gather features of what we would like and then see which of the vocabularies fits this best. My
guess is that we should perhaps move to schema.org over time, as that has a lot more useful features. Actually I opened an [issue 23 on solid-spec](https://github.com/solid/solid-spec/issues/23) on this a while ago. Still I think foaf is one of the simplest starting points for people to get going.

**Comment by sandhawke** [comment]
So we're branching off into a meta-issue.    I don't have it clear enough to raise it on its own, but maybe soon.    It's like HTML validity or XML well-formedness.   And it relates to Postel's Law.  In the general case, it might be phrased as "are there limits to Postel's Law?"   With HTML, it's: how much repair work should browsers do to bad HTML?    Before WHAT-WG, this was a nightmare, with one side saying "everyone must write valid HTML" and the other side saying "they don't" so the browsers each implemented different repair algorithms, and sites rendered badly.   As I understand it, the WHAT-WG solution was a middle ground that could be seen as standardizing the repair algorithm.  It could also be seen as expanding the language so that it's hard to make HTML that's outside the language.

In RDF we have this with some of the tools which consume schema.org markup doing lots of repair work.   

It's a slippery slope, or a race to the bottom.   Once the consuming systems start to be unpredictable about what kinds of repairs they'll do, producing systems start to be lax about what they emit, in part because they have no idea what will actually work.

The XML folks went the other way on this, saying XML parsers MUST reject XML documents that aren't well-formed, instead of repairing them.    There are tradeoffs.

The "try these four different predicates in order" is several steps down the slope toward clients being told they might as well do whatever they can to figure out a name for the user.    Maybe people suggesting it don't mean it that way.   Maybe they mean "do exactly these four steps in exactly this order, and nothing else".

I'd be -1 on racing to the bottom.   In the strongest possible terms.    It seems clear to me it would absolutely doom the project,

I'd be -0.9 on just-these-four-terms-and-never-anything-else because it seems like such an ugly waste.

To approach this from a slightly different angle, in terms of issue #72, you could view this as a sea of different possible standards for how to solve a particular problem.   Like, how many RDF syntaxes are we going to say that every client has to be able to parse?   Hopefully 1.

Okay, I think I know how to open that as its own issue now...

**Comment by timbl** [comment]
"We can gather features of what we would like and then see which of the vocabularies fits this best. "  No, not really.  Coming from the tabulator project we have many years of experience already.  For solid, we need to nail down an things like this without bikeshedding. Here is the code from a Labeller service in tabulator:

```
    var ns = tabulator.ns
    this.addLabelProperty(ns.vcard('fn'),10);
    this.addLabelProperty(ns.foaf('name'),10);
    this.addLabelProperty(ns.dct('title'),8);
    this.addLabelProperty(ns.dc('title'),7);
    this.addLabelProperty(ns.rss('title'),6);   // = dc:title?
    this.addLabelProperty(ns.contact('fullName'),4);
    this.addLabelProperty(kb.sym('http://www.w3.org/2001/04/roadmap/org#name'),4);
    this.addLabelProperty(ns.foaf('nick'),3);
    this.addLabelProperty(ns.rdfs('label'),2);
```
plus it adds **any subproperty of rdf:label** which it discovers.  That code is in https://linkeddata.github.io/tabulator/js/tab/labeler.js

((Note the priorities above give equal weight to foaf:name and vcard:fn 
I would be tempted to declared them equivalent properties.
I'm torn about which to promote, maybe foaf:name everywhere except with other vcard data or something.  Happy to treat them as the same))

**Comment by bblfish** [comment]
ah good, documenting existing implementations is also of course really helpful. It's not always clear where to look.

**Comment by dmitrizagidulin** [comment]
> I'm torn about which to promote

@timbl - let's promote vcard:fn and schema:name over foaf:name, then :)
(And yes, fall back on those other ones you have listed, when not there).

**Comment by csarven** [comment]
For whatever it is of worth, at the moment, dokieli checks for `foaf:name` with fallback to `schema:name`. I'd be comfortable enough to extend that to handle other common vocab terms I suppose. Essentially hardcoding some equivalences.

Apps need to be smart about what they can or want to consume. If they are only looking for `foaf:name`, they are going to mangle up when something else comes up. What this tells me that, when data is created by someone or through an application, they use the terms that's most suitable to their world-view, and adding equivalent terms to the data has a cost - whether that helps it to be reusable or not.

If we want to rely on any particular term for interoperability or what have you, then go with `rdfs:label` because that's generic as it gets, and the least one can do for any arbitrary resource. That's a "best practice", and from that you can get interoperability if that is indeed the primary goal here.

**Comment by deiu** [comment]
I'm personally against using `vcard:fn` since vcard is not meant to be a vocabulary to describe **yourself**. It's used to describe your personal view of **other** people.

My proposal is to use `foaf:name` and fall back on `schema:name`.

**Comment by dmitrizagidulin** [comment]
Ooh, that's not a bad idea, re `rdfs:label`..

**Comment by dmitrizagidulin** [comment]
@deiu - describing other people is part of the use case for this issue.

**Comment by deiu** [comment]
I thought the issue started from displaying our name when we log into an app/site.

**Comment by nicola** [comment]
As long as we could suggest, I see the following happening,

- Consider a profile to have only `vocab1:displayName`
- `App1` will ask for `vocab1:displayName`, because they just want a name to display, not the real name
- `App2` will ask for `vocab2:name` because they are interested their real name
- `App3` doesn't care about whether it is a nickname or a name+surname, they just want whatever the user has in its own profile amongst `vocab1:displayName`, `vocab2:name`, `vocab3:label`

In the case of `App2` cannot find property that they were looking for. Since `App2` is interested in knowing the actual name, they can prompt the user to add that new field in their profile.

In the case of `App3` cannot find the first property they were looking for, they will just fall back.

In other words, regardless of what we decide, applications are going to have the behavior of `App2` and `App3` regardless.

Of course it is important to suggest a particular property for guidance.

**Comment by sandhawke** [comment]
Why not pick one property and say that's what solid (at some version) uses?    And then have all our apps just set/use that one property....   Seems very straightforward to me.

I know lots of existing profiles won't have the property, sure.   But almost nothing else in the stack will work for them, either.    If the people-picker has fallback vocab, I can pick someone with a non-solid profile, but then when I grant them access to something, they won't be able to use that access, or when I try to put something in their inbox, it will fail.

I don't see much value in tiny bits of interop.    It's enough to get people thinking things might work, but they still won't actually work.    Is that somehow better than just having it clear upfront which pods use solid and which don't?

**Comment by kidehen** [comment]
@sandhawke What you are suggesting doesn't work and has never worked in this realm of RDF related development work. Net effect, we have an explosion of labels for a basic relation i.e, an annotation property that provides a label. 

rdf:label, foaf:name, schema:name, vcard:name, solid:name, {next-rdf-project}:name etc...

Why????

**Comment by sandhawke** [comment]
@kidehen Has the alternative worked either?   Where is there a set of interoperable apps, freely used by people, which follows this "everybody does a little bit of inference" approach?

**Comment by dmitrizagidulin** [comment]
>  Where is there a set of interoperable apps, freely used by people, which follows this "everybody does a little bit of inference" approach?

@sandhawke you are describing *the entirety of the Web* :)

**Comment by sandhawke** [comment]
Uhhhhhhhhhhhhh.      Interoperable?

**Comment by dmitrizagidulin** [comment]
.. Yes, inter-operable :) We're conversing, on this very issue, over hundreds of interoperable layers that all do a little bit of inference.

**Comment by kidehen** [comment]
@sandhawke -- if, then, else, is a basic reasoning construct. I think folks do that all the time when programming declaratively or imperatively. Having a countless number of annotation properties per application will never work. Either do RDF or don't do RDF. It is tiresome arguing about constructs that should be leveraging what RDF provides. 

If you want to design and implement an application that doesn't leverage features offered by standards such as RDF that's fine. Let's simply be clear about the nature of the application rather that trying use said application to redefine the nature of key standards like RDF.

**Comment by sandhawke** [comment]
I give up.   Y'all be crazy.

**Comment by kidehen** [comment]
@sandhawke : I am certainly crazy enough to believe in the use of open standards for loose-coupling of viewers, controllers, data models, and data sources. 

Very proud of that fact (outlined above) which sits at the core of any RDF related application we've developed for Web deployment to date. We know it works because we've written and deployed applications that leverage and demonstrate RDF utility. 

The very same thing will happen to SoLiD apps now that interop amongst core developers is gathering momentum. Stay tuned !

**Comment by melvincarvalho** [comment]
In my code I use:

```JavaScript
var FACE  = $rdf.Namespace("https://graph.facebook.com/schema/~/");
var FOAF  = $rdf.Namespace("http://xmlns.com/foaf/0.1/");

g.any(kb.sym(webid), FOAF('name'))  || g.any(kb.sym(webid), FACE('name')) ;
```

Rationale:

* foaf is used as we have been iterating on this in social linked data for 12 years.  See this detailed post from 2004 where major players wanted to adopt FOAF but ran in to issues around access control.  Issues that Solid today has solved.

* http://blog.plaxo.com/2004/02/plaxo_and_foaf_/

* I use facebook because it gives access to around a billion users / webids, which enables growth of the network.

It would be nice to see a developer best practice guide, but putting things like this in the core spec seems a little centralized.

**Comment by bblfish** [comment]
>  seems a little centralized.

It's not centralised it's one point of view. We should think of this as empirical research: list what is out there, what apps use, and that should help create reasons for coming together.

I point folks interested in the game theoretical underpinnings of convention to David Lewi's 1969 Thesis [Convention](http://www.princeton.edu/~harman/Courses/PHI534-2012-13/Nov26/lewis-convention1.pdf)

**Comment by dmitrizagidulin** [comment]
Further conversation on this on this morning's call with the QCRI team.
The consensus so far is:

1. Encourage (but not require) the use of `foaf:name` in the WebID profile. This is what apps should use to determine a user's name.
2. Specifically, developers of signup apps (such as [solid-signup](https://github.com/solid/solid-signup/)) should prompt the user for an optional name, and put that value into `foaf:name`
3. If `foaf:name` is absent, the app can always fall back on using the WebID URL itself as the username.
4. Falling back to other RDF properties (nick, schema:name, etc) is left to the discretion of app developers, and is outside the scope of the solid spec.

**Comment by akuckartz** [comment]
Meta-comment: This issue again proves that some of the most interesting technical discussions take place in issue trackers :smile: Maybe such issues should be labeled "Contains interesting discussion" so that they still can be found after they are closed.

**Comment by deiu** [comment]
Excellent point, @akuckartz!

**Comment by kidehen** [comment]
@akuckartz -- That's why we have nanotation. Basically, you can make notes using RDF Language sentences :) Example: 

{
<#thisKingsleyGithubComment2016-02-22-1>
a schema:Comment ;
schema:name "Kingsley Idehen Gitub Comment - 2016-02-22 #1" ;
schema:description """Use of Nanotation to do store fine-grained notes about this post.""";
schema:text """@akuckartz -- That's why we have nanotation. Basically, you can make notes using RDF Language sentences.""";
schema:mentions `<https://github.com/akuckartz#this>` ;
schema:mainEntity '<https://github.com/solid/solid/issues/74#this>' ;
schema:mainEntityOfPage <#this> .
}

**Comment by kidehen** [comment]
@dmitrizagidulin -- At juncture, all WebID+TLS and WebID+Profile implementations use foaf:name. Thus, it has to be treated as the key relation to which others are mapped. The mapping in question isn't a big deal anyway, as various comments have already indicated .

So :+1:  to your summary, in a nutshell :)

**Comment by dmitrizagidulin** [comment]
Added the naming recommendation to the spec: [Recommendation for User Names in Profiles](https://github.com/solid/solid-spec/blob/master/solid-webid-profiles.md#recommendation-for-user-names-in-profiles). Closing issue.
